sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,148 @@
+import { PollOperation } from "./pollOperation";
+
+export type CancelOnProgress = () => void;
+
+export interface PollProgressSubscriber<TProperties, TResult> {
+  id: string;
+  conditional: (op?: PollOperation<TProperties, TResult>) => boolean;
+  callback: (op?: PollOperation<TProperties, TResult>) => void;
+}
+
+export abstract class Poller<TProperties, TResult> {
+  private stopped: boolean;
+  private resolve?: (value?: TResult) => void;
+  private reject?: (error: Error) => void;
+  private pollOncePromise?: Promise<void>;
+  private cancelPromise?: Promise<PollOperation<TProperties, TResult>>;
+  private promise: Promise<TResult>;
+  private pollProgressSubscribers: PollProgressSubscriber<TProperties, TResult>[] = [];
+  public operation: PollOperation<TProperties, TResult>;
+
+  constructor(operation: PollOperation<TProperties, TResult>, stopped: boolean = false) {
+    this.operation = operation;
+    this.stopped = stopped;
+    this.promise = new Promise(
+      (
+        resolve: (result?: TResult) => void,
+        reject: (error: Error) => void
+      ) => {
+        this.resolve = resolve;
+        this.reject = reject;
+      }
+    );
+    this.operation.state.started = true;
+    this.startPolling();
+  }
+
+  abstract async delay(): Promise<void>;
+  abstract async getResult(): Promise<TResult | undefined>;
+
+  private async startPolling(): Promise<void> {
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(): Promise<void> {
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update();
+        if (this.isDone() && this.resolve) {
+          const result = await this.getResult();
+          this.operation.state.result = result;
+          this.resolve(result);
+        }
+      }
+    } catch (e) {
+      this.operation.state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+    }
+  }
+
+  public poll(): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce();
+      this.pollOncePromise.finally(() => {
+        for (const subscriber of this.pollProgressSubscribers) {
+          if (subscriber.conditional(this.operation)) {
+            subscriber.callback(this.operation);
+          }
+        }
+        this.pollOncePromise = undefined;
+      });
+    }
+    return this.pollOncePromise;
+  }
+
+  public done(): Promise<TResult | undefined> {
+    return this.promise;
+  }
+
+  public onProgress(
+    conditional: (op?: PollOperation<TProperties, TResult>) => boolean,
+    callback: (op?: PollOperation<TProperties, TResult>) => void
+  ): CancelOnProgress {
+    const id = Math.random().toString();
```

Comment:
this isn't a great source of random, better to use uuid() if you want to do this, but I don't think we need ids for this, because we can use object identity instead